src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -52,6 +52,12 @@ public override void ReadEnumerable(JsonTokenType tokenType, ref ReadStack state
         {
             Debug.Assert(ShouldDeserialize);
 
+            if (state.Current.KeyName == null && (state.Current.IsProcessingDictionary || state.Current.IsProcessingImmutableDictionary))
```

Comment:
It would be good to add a comment here explaining the condition, since its non-trivial.